core/src/main/scala/org/apache/spark/shuffle/ShuffleBlockFetcherIterable.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import java.lang.{Iterable => JIterable}
+import java.util.{Iterator => JIterator}
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.{ShuffleDependency, SparkConf, TaskContext}
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.serializer.SerializerManager
+import org.apache.spark.shuffle.api.io.ShuffleBlockInputStream
+import org.apache.spark.shuffle.api.metadata.ShuffleBlockInfo
+import org.apache.spark.shuffle.sort.io.LocalDiskShuffleBlockMetadata
+import org.apache.spark.storage.{BlockId, BlockManager, ShuffleBlockFetcherIterator, ShuffleBlockId}
+
+private[spark] class ShuffleBlockFetcherIterable(
+    blocks: JIterable[ShuffleBlockInfo],
+    context: TaskContext,
+    blockManager: BlockManager,
+    serializerManager: SerializerManager,
+    conf: SparkConf,
+    shouldBatchFetch: Boolean,
+    dep: ShuffleDependency[_, _, _])
+    extends JIterable[ShuffleBlockInputStream] with Logging {
+
+  private def fetchContinuousBlocksInBatch: Boolean = {
+    val serializerRelocatable = dep.serializer.supportsRelocationOfSerializedObjects
```

Comment:
This is the reason why we need to pass `ShuffleDependency` to the `ShuffleExecutorComponents#getPartitionReaders` API, but it seems less than ideal.